engine/common/handler/handler.go

```diff
@@ -0,0 +1,127 @@
+package handler
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// EventProcessorFunc is a function that processes an event. It is used by the AsyncEventHandler to process events.
+// The first argument is the identifier of the event's origin, and the second argument is the event itself.
+// The function is expected to be concurrency safe.
+// Processing an event is the last step in the event's lifecycle within the AsyncEventHandler.
+type EventProcessorFunc func(flow.Identifier, interface{})
+
+// AsyncEventHandler is a component that asynchronously handles events, i.e., concurrency safe and non-blocking.
+// It queues up the events and spawns a number of workers to process the events.
+type AsyncEventHandler struct {
```

Comment:
In my mind, this is essentially a worker pool with the following specific aspects: 
* processing function is already determined at construction time
* the queuing system for the worker pool is injected

_Suggestion_
* I think this implementation would benefit from generics. 
   * So instead of [`originId flow.Identifier, event interface{}`](https://github.com/onflow/flow-go/blob/21eccf36816a781e6ab29b8ef9cc8233e95ea799/engine/common/handler/handler.go#L113) we would use a generic type `T`
   * Note that you currently don't use `originId` (If `originId` is desired, it can be easily included as an implementation detail in `T`)
   * Your implementation looses type information as `event` is a general `interface`, requiring unnecessary casts in the code using the `AsyncEventHandler`

Reply:
https://github.com/onflow/flow-go/pull/3911/commits/41118f3f4635c57bc97372083336ff242d72ebc5